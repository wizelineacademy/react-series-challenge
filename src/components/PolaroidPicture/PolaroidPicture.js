import React, { Component } from 'react';
import {
  PolaroidListItem,
  PolaroidLink,
  PolaroidStar,
  PolaroidButton,
  PolaroidLoading,
} from './style.js'

class PolaroidPicture extends Component {

  constructor(props) {
    super(props);
    this.state = {
      loading: true
    };
  }

  render () {
    const { gif, favorites, favoriteGif, isSingleView, getGif } = this.props;
    if (!gif) {
      return null;
    }

    const handleImageChange = () => {
      this.setState({
        loading: false
      });
    };

    const isFavorite = favorites.includes(gif.id)
    
    return (
      <PolaroidListItem isSingleView={isSingleView}>
        <img src={gif.images.original.url} alt={gif.title} onLoad={handleImageChange.bind(this)} />
        { this.state.loading && <PolaroidLoading /> }
        <br />
        { !isSingleView && 
          <PolaroidButton className='like-button' onClick={() => favoriteGif(gif.id)} color={'red'}>
            { !isFavorite ? 'Like' : 'Unlike' }
          </PolaroidButton>
        }
        { !isSingleView && 
          <PolaroidLink to={`/gifs/${gif.id}`}>
            <PolaroidButton className='view-button' onClick={() => getGif(gif.id)} color={'blue'}>
              View
            </PolaroidButton>
          </PolaroidLink>
        }
        { isFavorite && <PolaroidStar /> }
        { isSingleView &&
          <p>
            Title: { gif.title } <br/>
            Created on: { gif.import_datetime } <br/>
            Short url: { gif.bitly_url } <br />
            SFW?: { gif.rating !== 'r' ? 'Yes!' : 'No! Why did you search for this?' } <br />
          </p>
        }
      </PolaroidListItem>
    )
  }
}

export default PolaroidPicture;