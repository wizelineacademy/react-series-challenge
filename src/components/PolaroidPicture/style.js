import styled from 'styled-components'
import { Link } from 'react-router-dom';


export const PolaroidListItem = styled.li`
  position: relative;
  width: ${props => props.isSingleView ? '500px;' : '200px;'}
  height: ${props => props.isSingleView ? '650px;' : '250px;'}
  display: inline-block;
  padding: 10px;
  margin: 10px;
  background-color: white;
  box-shadow: 7px 7px 4px black;
  border: 1px solid #000
  img {
    max-width: 100%;
    height: 80%;
  }
`
export const PolaroidLink = styled(Link)`
  color: white;
`

export const PolaroidStar = styled.div`
  background-image: url('/favorited.ico');
  background-size: contain;
  width: 50px;
  height: 50px;
  position: absolute;
  top: -25px;
  right: calc(50% - 25px);
`

export const PolaroidButton = styled.button`
  margin: 5px;
  padding: 10px 19px;
  margin: 0px 10px -10px 0px;
  border-radius: 10px;
  font-family: 'Pacifico', cursive;
  font-size: 17px;
  color: #FFF;
  text-decoration: none; 
  border: none;
  outline:none;
  cursor: pointer;
  display: none;
  ${ props => props.color === 'blue' ?
      `
        background-color: #3498DB;
        text-shadow: 0px -2px #2980B9;
      `
      :
      `
        background-color: #82BF56;
        text-shadow: 0px -2px #669644;
        float: right;
      `
  }
  :active {
    transform: translate(0px,5px);
    -webkit-transform: translate(0px,5px);
    border-bottom: 1px solid;
  }
`

export const PolaroidLoading = styled.span`
  position: absolute;
  width: 200px;
  height: 200px;
  left: 10px;
  background-image: url("/loader.svg");
  background-size: contain;
  background-color: black;
`