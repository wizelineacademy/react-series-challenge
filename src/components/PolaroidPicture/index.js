import PolaroidPicture from './PolaroidPicture';

export default PolaroidPicture