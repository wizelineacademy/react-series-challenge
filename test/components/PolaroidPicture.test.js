import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { MemoryRouter as Router } from 'react-router-dom';
import PolaroidPicture from '../../src/components/PolaroidPicture';

const IMAGE_URL = 'https://i.ytimg.com/vi/m5d1FlSeF-M/maxresdefault.jpg';
const IMAGE_ID = 'some_id';
const IMAGE_TITLE = 'some_title';

describe('PolaroidPicture', () => {
  let props;
  beforeEach(() => {
    props = {
      gif: {
        images: { original: { url: IMAGE_URL } },
        id: IMAGE_ID,
        title: IMAGE_TITLE,
        rating: 'n',
      },
      favorites: [],
      favoriteGif: jest.fn(),
      isSingleView: false,
      getGif: jest.fn(),
    }
  })

  it('should render correctly', () => {
    const tree = renderer
    .create(
      <Router><PolaroidPicture {...props} /></Router>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should call favoriteGif callback when like button is clicked', () => {
    const output = shallow(
      <PolaroidPicture {...props} />
    );
    expect(props.favoriteGif.mock.calls.length).toBe(0);
    output.find('.like-button').simulate('click');
    expect(props.favoriteGif.mock.calls.length).toBe(1);
  });

  it('should call getGif callback when view button is clicked', () => {
    const output = shallow(
      <PolaroidPicture {...props} />
    );
    expect(props.getGif.mock.calls.length).toBe(0);
    output.find('.view-button').simulate('click');
    expect(props.getGif.mock.calls.length).toBe(1);
  });

  it('should call favoriteGif callback when like button is clicked', () => {
    const output = shallow(
      <PolaroidPicture {...props} />
    );
    expect(props.favoriteGif.mock.calls.length).toBe(0);
    output.find('.like-button').simulate('click');
    expect(props.favoriteGif.mock.calls.length).toBe(1);
  });

  it('should not render if gif is null', () => {
    const tree = renderer
    .create(
      <PolaroidPicture gif={null} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render appropiately if isSingleView', () => {
    props.isSingleView = true;
    const tree = renderer
    .create(
      <PolaroidPicture gif={null} />
    ).toJSON();

    expect(tree).toMatchSnapshot();    
  });

  it('should have easter egg if gif rating is r and isSingleView', () => {
    props.isSingleView = true;
    props.gif.rating = 'r';
    const tree = renderer
    .create(
      <PolaroidPicture gif={null} />
    ).toJSON();

    expect(tree).toMatchSnapshot();    
  });

});