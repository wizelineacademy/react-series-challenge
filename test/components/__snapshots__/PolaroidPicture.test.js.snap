// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolaroidPicture should have easter egg if gif rating is r and isSingleView 1`] = `null`;

exports[`PolaroidPicture should not render if gif is null 1`] = `null`;

exports[`PolaroidPicture should render appropiately if isSingleView 1`] = `null`;

exports[`PolaroidPicture should render correctly 1`] = `
.c0 {
  position: relative;
  width: 200px;
  height: 250px;
  display: inline-block;
  padding: 10px;
  margin: 10px;
  background-color: white;
  box-shadow: 7px 7px 4px black;
  border: 1px solid #000;
}

.c0 img {
  max-width: 100%;
  height: 80%;
}

.c3 {
  color: white;
}

.c2 {
  margin: 5px;
  padding: 10px 19px;
  margin: 0px 10px -10px 0px;
  border-radius: 10px;
  font-family: 'Pacifico',cursive;
  font-size: 17px;
  color: #FFF;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: none;
  outline: none;
  cursor: pointer;
  display: none;
  background-color: #82BF56;
  text-shadow: 0px -2px #669644;
  float: right;
}

.c2:active {
  -webkit-transform: translate(0px,5px);
  -ms-transform: translate(0px,5px);
  transform: translate(0px,5px);
  -webkit-transform: translate(0px,5px);
  border-bottom: 1px solid;
}

.c4 {
  margin: 5px;
  padding: 10px 19px;
  margin: 0px 10px -10px 0px;
  border-radius: 10px;
  font-family: 'Pacifico',cursive;
  font-size: 17px;
  color: #FFF;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: none;
  outline: none;
  cursor: pointer;
  display: none;
  background-color: #3498DB;
  text-shadow: 0px -2px #2980B9;
}

.c4:active {
  -webkit-transform: translate(0px,5px);
  -ms-transform: translate(0px,5px);
  transform: translate(0px,5px);
  -webkit-transform: translate(0px,5px);
  border-bottom: 1px solid;
}

.c1 {
  position: absolute;
  width: 200px;
  height: 200px;
  left: 10px;
  background-image: url("/loader.svg");
  background-size: contain;
  background-color: black;
}

<li
  className="c0"
>
  <img
    alt="some_title"
    onLoad={[Function]}
    src="https://i.ytimg.com/vi/m5d1FlSeF-M/maxresdefault.jpg"
  />
  <span
    className="c1"
  />
  <br />
  <button
    className="like-button c2"
    color="red"
    onClick={[Function]}
  >
    Like
  </button>
  <a
    className="c3"
    href="/gifs/some_id"
    onClick={[Function]}
  >
    <button
      className="view-button c4"
      color="blue"
      onClick={[Function]}
    >
      View
    </button>
  </a>
</li>
`;
